### ISSUE TITLE
[js] enum-as-object cleanup wishlist

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#7165]
So I diffed the output with the new object-based enums and it looks great, however it increases code size in real-world cases quite a bit and I think we can do something about it:

```haxe
enum E {
	A;
	B(v:Int);
}
```

generates

```js
var E = { __ename__ : true, __constructs__ : ["A","B"] };
$hxEnums["E"] = E;
E.A = {_hx_index:0};
E.A.toString = $estr;
E.A.__enum__ = "E";
E.B = function(v) { var $x = {_hx_index:1,v:v,__enum__:"E"}; $x.toString = $estr; return $x; }
E.B.__params__ = ["v"];;
```

could be optimized to

```js
var E = $hxEnums["E"] = { __ename__ : true, __constructs__ : ["A","B"] };
E.A = {_hx_index:0, toString: $estr, __enum__: "E"};
E.B = ($_ = function(v) { return {_hx_index:1,v:v,__enum__:"E",toString: $estr} }, $_.__params__ = ["v"], $_);
```

Basically we don't want to repat the type/ctor name too much because in real code it would be something like `rambo_ResourceChangeReason.ClanwarAttackMaterialPurchase`.

---

Another thing that makes me uncomfortable is the naming inconsistency here. We have `_hx_index`, but also `__enum__` and `$hxEnums`. I feel like it would be better to use the `$ident` scheme for everything as this would also exclude the possibility of name clashes coming from weirdly-named identifiers in Haxe code.

So something like `E.A = {$enum: "E", $index: 0, toString: $estr};`. I think it would also look better when using the enum objects from plain JavaScript, as the `$`-fields look like special tag fields while non-`$`-fields contain data.

---

Thoughts?

**Comment by ncannasse** [comment]
Yes $ prefix is a better escaping sequence for JS output since it's not a valid Haxe identifier

**Comment by nadako** [comment]
Another thing - since we have `$hxEnums` now, it makes sense to make `Type.resolveEnum` use it instead of also putting enums into `$hxClasses`.

**Comment by nadako** [comment]
It's better now:

```js
var E = $hxEnums["E"] = { __ename__ : true, __constructs__ : ["A","B"] };
E.A = {_hx_index:0,__enum__:"E",toString:$estr};
E.B = ($_=function(v) { return {_hx_index:1,v:v,__enum__:"E",toString:$estr}; },$_.__params__ = ["v"],$_);
```

still need to rename the fields (`$enum` and `$index`?)

another idea is to merge the declarations into one object, so we don't repeat the type name at all:
```js
var E = {
  __constructs__: ["A", "B"],
  A: {$index: 0, $enum: "E"},
  B: ($_=function(v) { return {$index: 1, $enum: "E", v:v} },$_.__params__=["v"],$_),
}
```

**Comment by ncannasse** [comment]
$enum and $index is good for me. and yes the litteral declaration seems better

**Comment by Simn** [comment]
Do you still want to do something here for 4.0?

**Comment by nadako** [comment]
I wanted to unify impl fields to use `$` instead of double underscores and `_hx_` prefixes, but it's a bit annoying to implement because we want to access those fields from haxe code (`Type.enumIndex` etc). Not a priority for 4.0, I'd say.

**Comment by nadako** [comment]
I still think we should use `$`-prefixed fields, but it can't be done in a non-ugly way without #5105 or #9433 in place, because, as mentioned, we need to access them in the reflection API.

**Comment by nadako** [comment]
Also I don't like the fact that `__enum__` is a string. IMO it should cycle-reference the enum definition object and the JSON round-trip argument seem weak to me (direct `JSON.stringify` on enums is not something anyone would want to do anyway).

So if there are no arguments against changing this, I'll do this at some point when I have time for it :)
